File: app/api/routers/ticket.py
from fastapi import APIRouter, Depends, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.database import get_async_session
from app.schemas import (
    TicketCreate,
    TicketListParams,
    TicketListResponse,
    TicketRead,
    TicketStatusUpdate,
)
from app.services import TicketService
from app.core.JWT.admin_auth import require_admin

ticket_router  = APIRouter(
    prefix="/tickets", 
    tags=["tickets"]
)

@ticket_router.get('', response_model=TicketListResponse)
async def get_tickets(
    params: TicketListParams = Depends(),
    session: AsyncSession = Depends(get_async_session)
):
    return await TicketService.get_list(session=session, params=params)

@ticket_router.get("/{id}", response_model=TicketRead)
async def get_ticket_by_id(
    id: int,
    session: AsyncSession = Depends(get_async_session),
):
    ticket = await TicketService.get_by_id(session=session, id=id)
    return TicketRead.model_validate(ticket)

@ticket_router.post(
        '', 
        status_code=status.HTTP_201_CREATED,
        response_model=TicketRead
)
async def create(
    data: TicketCreate,
    session: AsyncSession = Depends(get_async_session)
):
    ticket = await TicketService.create(session=session, data=data)
    return TicketRead.model_validate(ticket)

@ticket_router.patch("/{id}/status", response_model=TicketRead)
async def update_ticket_status(
    id: int,
    data: TicketStatusUpdate,
    session: AsyncSession = Depends(get_async_session),
):
    ticket = await TicketService.update_status(session=session, id=id, data=data)
    return TicketRead.model_validate(ticket)

@ticket_router.delete("/{id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_ticket(
    id: int,
    session: AsyncSession = Depends(get_async_session),
    current_admin = Depends(require_admin)
):
    await TicketService.delete(session=session, id=id)

File: frontendReact/src/api/ticket.ts
import { apiUrl } from "./config";
import { authFetch } from "./authFetch";

import type {
  Ticket,
  TicketCreate,
  TicketListParams,
  TicketListResponse,
  TicketStatus,
} from "../types/ticket";

function buildTicketQuery(params: TicketListParams = {}): string {
  const searchParams = new URLSearchParams();

  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== null && value !== "") {
      searchParams.set(key, String(value));
    }
  });

  const query = searchParams.toString();

  return query ? `?${query}` : "";
}

export async function getTickets(
  params: TicketListParams = {},
): Promise<TicketListResponse> {
  const query = buildTicketQuery(params);

  return authFetch<TicketListResponse>(apiUrl(`/tickets${query}`));
}

export async function createTicket(data: TicketCreate): Promise<Ticket> {
  return authFetch<Ticket>(apiUrl("/tickets"), {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export async function updateTicketStatus(
  id: number,
  status: TicketStatus,
): Promise<Ticket> {
  return authFetch<Ticket>(apiUrl(`/tickets/${id}/status`), {
    method: "PATCH",
    body: JSON.stringify({ status }),
  });
}

export async function deleteTicket(id: number): Promise<void> {
  return authFetch<void>(apiUrl(`/tickets/${id}`), {
    method: "DELETE",
  });
}
